File: ConduitCore/src/main/java/systems/conduit/core/mixins/ServerGamePacketListenerMixin.java
package systems.conduit.core.mixins;

import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.network.ServerGamePacketListenerImpl;
import net.minecraft.server.players.PlayerList;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.ModifyArg;
import org.spongepowered.asm.mixin.injection.Redirect;
import systems.conduit.core.Conduit;
import systems.conduit.core.events.types.PlayerEvents;

@Mixin(value = ServerGamePacketListenerImpl.class, remap = false)
public class ServerGamePacketListenerMixin {

    @Shadow public ServerPlayer player;

    @Redirect(method = "handleChat", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/players/PlayerList;broadcastMessage(Lnet/minecraft/network/chat/Component;Z)V"))
    private void handleChat(PlayerList playerList, Component component, boolean b) {
        PlayerEvents.PlayerChatEvent event = new PlayerEvents.PlayerChatEvent((systems.conduit.api.Player) this.player, component);
        Conduit.getEventManager().dispatchEvent(event);
        Component eventMessage = event.getMessage();
        if (eventMessage != null) playerList.broadcastMessage(eventMessage, b);
    }

    @ModifyArg(method = "handleChat", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/network/ServerGamePacketListenerImpl;handleCommand(Ljava/lang/String;)V"))
    private String handleCommand(String message) {
        PlayerEvents.PlayerCommandEvent event = new PlayerEvents.PlayerCommandEvent((systems.conduit.api.Player) this.player, message);
        Conduit.getEventManager().dispatchEvent(event);
        return event.getMessage();
    }

    @Redirect(method = "onDisconnect", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/players/PlayerList;broadcastMessage(Lnet/minecraft/network/chat/Component;)V"))
    private void playerLeaveMessage(PlayerList playerList, Component message, Component arg) {
        PlayerEvents.LeaveType leaveType = PlayerEvents.LeaveType.UNKNOWN;
        if (arg instanceof TranslatableComponent) {
            TranslatableComponent type = ((TranslatableComponent) arg);
            if (type.getKey().equalsIgnoreCase("multiplayer.disconnect.kicked")) {
                leaveType = PlayerEvents.LeaveType.KICKED;
            } else if (type.getKey().equalsIgnoreCase("multiplayer.disconnect.generic")) {
                leaveType = PlayerEvents.LeaveType.LEFT;
            }
        }
        PlayerEvents.PlayerLeaveEvent event = new PlayerEvents.PlayerLeaveEvent((systems.conduit.api.ServerPlayer) player, message, leaveType);
        Conduit.getEventManager().dispatchEvent(event);
        Component eventMessage = event.getMessage();
        if (eventMessage != null) playerList.broadcastMessage(event.getMessage());
    }
}

File: src/main/kotlin/systems/conduit/core/mixins/ServerGamePacketListenerMixin.kt
package systems.conduit.core.mixins

import net.minecraft.network.chat.Component
import net.minecraft.network.chat.TextComponent
import net.minecraft.network.chat.TranslatableComponent
import net.minecraft.server.level.ServerPlayer
import net.minecraft.server.network.ServerGamePacketListenerImpl
import net.minecraft.server.players.PlayerList
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.ModifyArg
import org.spongepowered.asm.mixin.injection.Redirect
import systems.conduit.api.Player
import systems.conduit.core.Conduit
import systems.conduit.core.events.types.PlayerEvents

@Mixin(value = [ServerGamePacketListenerImpl::class], remap = false)
class ServerGamePacketListenerMixin {

    @Shadow var player: ServerPlayer? = null

    @Redirect(method = ["handleChat"], at = At(value = "INVOKE", target = "Lnet/minecraft/server/players/PlayerList;broadcastMessage(Lnet/minecraft/network/chat/Component;Z)V"))
    private fun handleChat(playerList: PlayerList?, component: Component?, b: Boolean) {
        val event = PlayerEvents.PlayerChatEvent(player as Player, component)
        Conduit.eventManager.dispatchEvent(event)
        if (playerList != null && component != null) playerList.broadcastMessage(event.message, b)
    }

    @ModifyArg(method = ["handleChat"], at = At(value = "INVOKE", target = "Lnet/minecraft/server/network/ServerGamePacketListenerImpl;handleCommand(Ljava/lang/String;)V"))
    private fun handleCommand(message: String?): String? {
        var eventMessage = message
        if (eventMessage == null) eventMessage = ""
        val event = PlayerEvents.PlayerCommandEvent(player as Player, eventMessage)
        Conduit.eventManager.dispatchEvent(event)
        return event.message
    }

    @Redirect(method = ["onDisconnect"], at = At(value = "INVOKE", target = "Lnet/minecraft/server/players/PlayerList;broadcastMessage(Lnet/minecraft/network/chat/Component;)V"))
    private fun playerLeaveMessage(playerList: PlayerList?, message: Component?, arg: Component?) {
        var leaveType: PlayerEvents.LeaveType = PlayerEvents.LeaveType.UNKNOWN
        if (arg is TranslatableComponent) {
            if (arg.key.equals("multiplayer.disconnect.kicked", ignoreCase = true)) {
                leaveType = PlayerEvents.LeaveType.KICKED
            } else if (arg.key.equals("multiplayer.disconnect.generic", ignoreCase = true)) {
                leaveType = PlayerEvents.LeaveType.LEFT
            }
        }
        val event = PlayerEvents.PlayerLeaveEvent(player as Player, message, leaveType)
        Conduit.eventManager.dispatchEvent(event)
        if (playerList != null && message != null) playerList.broadcastMessage(event.message)
    }
}
